scripts: check for control flow instrumentation in security-check.py

fanquake opened the pull request:
Note that the presence of instrumentation does not mean it will be used, as that is determined at runtime by the CPU.
From the Intel control flow enforcement documentation:
> The ENDBR32 and ENDBR64 instructions will have the same effect as the NOP instruction on Intel 64 processors that do not support CET. On processors supporting CET, these instructions do not change register or flag state. This allows CET instrumented programs to execute on processors that do not support CET. Even when CET is supported and enabled, these NOP–like instructions do not affect the execution state of the program, do not cause any additional register pressure, and are minimally intrusive from power and performance perspectives.

fanquake committed:
scripts: add separate_code failure to all calls in ELF test security check

fanquake commented:
Aside from the missing test additions, this approach is currently too naive. There is likely to be the presence of `endbr64` instructions in the .text section, which we do not control (i.e from libc). I'm also not sure that targeting a specific symbol via objdump will work, as that will still dump all output after the symbol (and may require a too newer version of objdump).

laanwj commented:
> There is likely to be the presence of endbr64 instructions in the .text section, which we do not control (i.e from libc). 

The check is naive, but maybe enough. Let's update test-security-check for it and see if it distinguishes between the hardening flag being present or not?

Remember that we do not link libc statically at this point so there should be no instructions from libc in the binary.

laanwj commented on contrib/devtools/security-check.py:
```diff
@@ -184,6 +198,14 @@ def check_PE_NX(executable) -> bool:
     return (bits & IMAGE_DLL_CHARACTERISTICS_NX_COMPAT) == IMAGE_DLL_CHARACTERISTICS_NX_COMPAT
 
+def check_PE_control_flow_instrumentation(executable) -> bool:
+    stdout = run_command([OBJDUMP_CMD, '-d', '--section=.text', executable])
```
Does PE have a `.text` section?

fanquake replied:
Yes, see https://docs.microsoft.com/en-us/windows/win32/debug/pe-format#special-sections, and filltering by --section here prevents additional output from `objdump`.

laanwj commented on contrib/devtools/security-check.py:
```diff
@@ -143,6 +143,20 @@ def check_ELF_separate_code(executable):
+    elf = pixie.load(executable)
+
+    if elf.hdr.e_machine is not pixie.EM_X86_64:
+        return True
```
At some point we might want to add a 'skipped' return status for checks that could not be performed for a platform (they would count as true for the final return value, but might be displayed differently).

fanquake replied:
Agree. I think printing some sort of informational output may be useful.

laanwj commented on contrib/devtools/security-check.py:
```diff
@@ -184,6 +198,14 @@ def check_PE_NX(executable) -> bool:
+    stdout = run_command([OBJDUMP_CMD, '-d', '--section=.text', executable])
+
+    for line in stdout.splitlines():
+        if 'endbr64' in line:
```
It's too bad that x86 has variable-length instructions (so it's tricky to find instruction binaries), or we could just look in the section contents for the raw instruction instead of parsing text.

fanquake replied:
You've implemented this!

prusnak commented on contrib/devtools/security-check.py:
```diff
@@ -257,12 +279,14 @@ def check_MACHO_Canary(executable) -> bool:
     ('RELRO', check_ELF_RELRO),
     ('Canary', check_ELF_Canary),
     ('separate_code', check_ELF_separate_code),
+    ('control_flow', check_ELF_control_flow_instrumentation)
```
```suggestion
    ('control_flow', check_ELF_control_flow_instrumentation),
```

fanquake replied:
Thanks. Included these suggestions.

prusnak commented on contrib/devtools/security-check.py:
```diff
@@ -257,12 +279,14 @@ def check_MACHO_Canary(executable) -> bool:
     ('NX', check_PE_NX),
-    ('RELOC_SECTION', check_PE_RELOC_SECTION)
+    ('RELOC_SECTION', check_PE_RELOC_SECTION),
+    ('control_flow', check_PE_control_flow_instrumentation)
```
```suggestion
    ('control_flow', check_PE_control_flow_instrumentation),
```

prusnak reviewed (changes requested):
Python supports trailing commas in lists. Using them avoids future changes unrelated to the desired change.

laanwj committed:
contrib: Implement ELF control_flow check with pixie, add test

Don't need to do shell out to `objdump` for ELF, we simply check if
`main` starts with the right instruction.

Add a test in `test-security-check.py`.

fanquake committed:
contrib: add control flow check for PE binaries

practicalswift commented:
Concept ACK: hardening is good and CFI is great! :)

fanquake commented:
I've pulled in @laanwj's changes, as well as added a basic test for the PE binaries. One thing to note is that our `test_bitcoin.exe` binary doesn't currently have an `endbr64` instruction in `main`, like all the other PE binaries.

The only `endbr64` instructions that come after `main` are from some ZeroMQ functions:
```bash
objdump -C --disassemble=main --section=.text src/test/test_bitcoin.exe | rg endbr64 -C5
 13c670d:	90                   	nop
 13c670e:	90                   	nop
 13c670f:	90                   	nop

00000000013c6710 <_GLOBAL__sub_I__ZN25CZMQNotificationInterfaceC2Ev>:
 13c6710:	f3 0f 1e fa          	endbr64 
 13c6714:	53                   	push   %rbx
 13c6715:	48 83 ec 30          	sub    $0x30,%rsp
 13c6719:	48 8b 1d 10 fd 1f 00 	mov    0x1ffd10(%rip),%rbx        # 15c6430 <.refptr.__stack_chk_guard>
 13c6720:	48 8d 15 c1 25 1f 00 	lea    0x1f25c1(%rip),%rdx        # 15b8ce8 <.rdata+0xb98>
 13c6727:	48 8d 0d f2 aa 37 00 	lea    0x37aaf2(%rip),%rcx        # 1741220 <CURRENCY_UNIT>
--
 13c679d:	90                   	nop
 13c679e:	90                   	nop
 13c679f:	90                   	nop

00000000013c67a0 <_GLOBAL__sub_I__ZN27CZMQAbstractPublishNotifier10InitializeEPv>:
 13c67a0:	f3 0f 1e fa          	endbr64 
 13c67a4:	53                   	push   %rbx
 13c67a5:	48 83 ec 30          	sub    $0x30,%rsp
 13c67a9:	48 8b 1d 80 fc 1f 00 	mov    0x1ffc80(%rip),%rbx        # 15c6430 <.refptr.__stack_chk_guard>
 13c67b0:	48 8d 0d 09 ab 37 00 	lea    0x37ab09(%rip),%rcx        # 17412c0 <std::__ioinit>
 13c67b7:	48 8b 03             	mov    (%rbx),%rax
--
 13c686d:	90                   	nop
 13c686e:	90                   	nop
 13c686f:	90                   	nop

00000000013c6870 <_GLOBAL__sub_I__Z22RegisterZMQRPCCommandsR9CRPCTable>:
 13c6870:	f3 0f 1e fa          	endbr64 
 13c6874:	41 54                	push   %r12
 13c6876:	53                   	push   %rbx
 13c6877:	48 83 ec 58          	sub    $0x58,%rsp
 13c687b:	48 8b 1d ae fb 1f 00 	mov    0x1ffbae(%rip),%rbx        # 15c6430 <.refptr.__stack_chk_guard>
 13c6882:	48 8d 15 57 37 1f 00 	lea    0x1f3757(%rip),%rdx        # 15b9fe0 <.rdata+0x560>
--
 13c6936:	48 83 c4 58          	add    $0x58,%rsp
 13c693a:	5b                   	pop    %rbx
 13c693b:	41 5c                	pop    %r12
 13c693d:	c3                   	retq   
 13c693e:	e8 9d ed cc ff       	callq  10956e0 <__stack_chk_fail>
 13c6943:	f3 0f 1e fa          	endbr64 
 13c6947:	48 8b 4c 24 20       	mov    0x20(%rsp),%rcx
 13c694c:	49 89 c4             	mov    %rax,%r12
 13c694f:	48 8d 44 24 30       	lea    0x30(%rsp),%rax
 13c6954:	48 39 c1             	cmp    %rax,%rcx
 13c6957:	74 05                	je     13c695e <_GLOBAL__sub_I__Z22RegisterZMQRPCCommandsR9CRPCTable+0xee>
--
 13c6a2c:	5d                   	pop    %rbp
 13c6a2d:	41 5c                	pop    %r12
 13c6a2f:	c3                   	retq   

00000000013c6a30 <_GLOBAL__sub_I__Z8zmqErrorPKc>:
 13c6a30:	f3 0f 1e fa          	endbr64 
 13c6a34:	53                   	push   %rbx
 13c6a35:	48 83 ec 30          	sub    $0x30,%rsp
 13c6a39:	48 8b 1d f0 f9 1f 00 	mov    0x1ff9f0(%rip),%rbx        # 15c6430 <.refptr.__stack_chk_guard>
 13c6a40:	48 8d 0d 79 a9 37 00 	lea    0x37a979(%rip),%rcx        # 17413c0 <std::__ioinit>
 13c6a47:	48 8b 03             	mov    (%rbx),%rax
```

There are more `endbr64` instructions in the binary:
```bash
objdump -d --section=.text src/test/test_bitcoin.exe | rg endbr64 | wc -l
59313
```

they are all just pre-main in the `.text` section, and main itself doesn't have one.

Looks like we also need to fix the linter for the pixie additions:
```bash
contrib/devtools/pixie.py:211: error: Item "None" of "Optional[Dict[int, bytes]]" has no attribute "get"
contrib/devtools/pixie.py:213: error: Incompatible types in assignment (expression has type "Iterator[None]", variable has type "Generator[Union[bytes, None, Any], None, None]")
Found 2 errors in 1 file (checked 192 source files)
^---- failure generated from test/lint/lint-python.sh
```

fanquake commented:
Another issue is that the [CI which currently runs the security check tests](https://github.com/bitcoin/bitcoin/blob/master/ci/test/00_setup_env_native_qt5.sh) uses Bionic, which has a GCC (7.4.0) which does not have the `-fcf-protection` option. So we'll need to move the running of the security check tests to one of the CIs using Focal.
```bash
gcc: error: unrecognized command line option ‘-fcf-protection=none’; did you mean ‘-flto-partition=none’?
E
======================================================================
ERROR: test_ELF (__main__.TestSecurityChecks)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "./contrib/devtools/test-security-check.py", line 34, in test_ELF
    self.assertEqual(call_security_check(cc, source, executable, ['-Wl,-zexecstack','-fno-stack-protector','-Wl,-znorelro','-no-pie','-fno-PIE', '-Wl,-z,noseparate-code', '-fcf-protection=none']),
  File "./contrib/devtools/test-security-check.py", line 23, in call_security_check
    subprocess.run([cc,source,'-o',executable] + options, check=True)
  File "/usr/lib/python3.6/subprocess.py", line 438, in run
    output=stdout, stderr=stderr)
subprocess.CalledProcessError: Command '['gcc', 'test1.c', '-o', 'test1', '-Wl,-zexecstack', '-fno-stack-protector', '-Wl,-znorelro', '-no-pie', '-fno-PIE', '-Wl,-z,noseparate-code', '-fcf-protection=none']' returned non-zero exit status 1.

----------------------------------------------------------------------
Ran 1 test in 0.046s

FAILED (errors=1)
```

laanwj commented:
Is there anything specific holding this up?

fanquake commented:
> Is there anything specific holding this up?

Just my lack of getting things done. I was planning on modifying this so we didn't have to special case `test_bitcoin.exe`

fanquake commented:
Going to close this PR, open an issue explaining why control flow instrumentation isn't always present, and then open separate PRs per platform modifying dependencies and adding the security checks. The way we test for this will also be simpler now that we are using LIEF.
